### PR TITLE
feat: price overide

### LINKED ISSUE #128: Enhancement: Price Override
As [I was cleaning up](https://github.com/ubiquity/ubiquity-dollar/issues/225#issuecomment-1432554966) the deprecated labels on the ubiquity-dollar repo, I realized that it would be nice for the bot to not fight me if I assign a `Price: ` label.

The bot should understand that if I set a Price label, to not wipe it. This also allows maintainers to make custom (financial) deals with bounty hunters as well if needed.

### PR DESCRIPTION
**Pull request by BeanieMen**
Resolves #128

### CODE REVIEW

**Comment by me505** [comment]
@rndquu this might work

**Comment by pavlovcik** [comment]
You need to post working examples of the expected results from your bot

**Comment by me505** [comment]
I can't qa this is tried but I am getting `Error logging to supabase: undefined` and a whole bunch of others

**Comment by me505** [comment]
@pavlovcik know any fixes for this

**Comment by pavlovcik** [comment]
@ubiquity/software-development

**Comment by wannacfuture** [comment]
Check if you have set supabase - related env variables.
Also, even it is saying like that, there will be no problem for the bot to act.

You can just ignore supabase loggin related errors. @me505

File: src/handlers/pricing/action.ts [hunk 1]
```diff
@@ -37,10 +37,44 @@ export const pricingLabelLogic = async (): Promise<void> => {
...
+        // get all issue events of type "labeled" and the event label includes Price
+        const events = await getAllIssueEvents();
+        if (!events) return;
+        const labeledEvents: typeof events = [];
+        const labeledPriceEvents: typeof events = [];
+        events.forEach((event) => {
+          if (event.event === "labeled") {
+            labeledEvents.push(event);
+          }
+        });
```
Comment (wannacfuture) [inline src/handlers/pricing/action.ts:53]: can we make this as a util function? like getLabeledEvents or sth. We need to make them as much as reusable.

**Review by wannacfuture — CHANGES_REQUESTED** [review]

**Comment by me505** [comment]
I have set the supabase url and key

File: src/handlers/pricing/action.ts (hunk 1 above)
Comment (me505) [reply src/handlers/pricing/action.ts:53]: After I come back from school will do

**Comment by me505** [comment]
in the issue i mentioned nothings seems to work

**Comment by whilefoo** [comment]
> I can't qa this is tried but I am getting `Error logging to supabase: undefined` and a whole bunch of others

make sure you run all migrations (https://github.com/ubiquity/ubiquibot/tree/development/supabase)
`yarn supabase db push`

**Comment by me505** [comment]
first is the oldest last is the newest kinda like a stack

**Comment by me505** [comment]
gotta fix that

**Comment by me505** [comment]
https://github.com/me505/ubiquibot/issues/6

**Comment by me505** [comment]
@whilefoo heres the qa

**Comment by me505** [comment]
https://github.com/me505/ubiquibot/issues/6

**Comment by me505** [comment]
it generates correct permit now

File: src/handlers/wildcard/analytics.ts [hunk 2]
```diff
@@ -27,7 +27,13 @@ export const bountyInfo = (
   const minTimeLabel = timeLabels.length > 0 ? timeLabels.reduce((a, b) => (calculateWeight(a) < calculateWeight(b) ? a : b)).name : undefined;
   const minPriorityLabel = priorityLabels.length > 0 ? priorityLabels.reduce((a, b) => (calculateWeight(a) < calculateWeight(b) ? a : b)).name : undefined;
 
-  const priceLabel = getTargetPriceLabel(minTimeLabel, minPriorityLabel);
+  const labelNames = labels.map((i) => i.name);
+  let priceLabel = "";
+  labelNames.forEach((e) => {
```
Comment (whilefoo) [inline src/handlers/wildcard/analytics.ts:32]: I think using `.find` would be more clear
```
const priceLabel = labels.find(label => label.name.includes("Price"))
```

File: src/handlers/pricing/action.ts [hunk 3]
```diff
@@ -37,10 +37,53 @@ export const pricingLabelLogic = async (): Promise<void> => {
...
+        const labeledEvents = await getAllLabeledEvents();
+        if (!labeledEvents) return;
+        const labeledPriceEvents: typeof labeledEvents = [];
+        labeledEvents.forEach((event) => {
+          if (event.label?.name.includes("Price")) {
+            labeledPriceEvents.push(event);
+          }
+        });
+        // check if the latest price label has been added by a user
+        if (labeledPriceEvents[labeledPriceEvents.length - 1].actor?.type == "User") {
```
Comment (whilefoo) [inline src/handlers/pricing/action.ts:69]: add check for array length and you can use `UserType` enum instead of `User` literal

File: src/handlers/pricing/action.ts [hunk 4]
```diff
@@ -37,10 +37,53 @@ export const pricingLabelLogic = async (): Promise<void> => {
...
+    });
+
+    if (hasPriceLabel) {
+      // below checks each label to see if theres a price
+      if (hasTargetPriceLabel) {
+        // get all issue events of type "labeled" and the event label includes Price
+        const labeledEvents = await getAllLabeledEvents();
+        if (!labeledEvents) return;
+        const labeledPriceEvents: typeof labeledEvents = [];
+        labeledEvents.forEach((event) => {
```
Comment (whilefoo) [inline src/handlers/pricing/action.ts:63]: you can use `labeledEvents.filter()`

File: src/helpers/issue.ts [hunk 5]
```diff
@@ -3,6 +3,37 @@ import { getBotConfig, getBotContext, getLogger } from "../bindings";
...
+  try {
+    // Fetch issue events
+    const { data: events } = await context.octokit.issues.listEvents({
+      owner: payload.repository.owner.login,
+      repo: payload.repository.full_name,
+      issue_number: payload.issue.number,
+    });
+    return events;
+  } catch (e: unknown) {
+    logger.debug(`Getting all issue events failed, reason: ${e}`);
```
Comment (whilefoo) [inline src/helpers/issue.ts:20]:
```suggestion
    logger.error(`Getting all issue events failed, reason: ${e}`);
```

File: src/helpers/issue.ts [hunk 6]
```diff
@@ -3,6 +3,37 @@ import { getBotConfig, getBotContext, getLogger } from "../bindings";
...
+    return null;
+  }
+};
+
+export const getAllLabeledEvents = async () => {
+  const events = await getAllIssueEvents();
+  if (!events) return;
+  const labeledEvents: typeof events = [];
+  events.forEach((event) => {
```
Comment (whilefoo) [inline src/helpers/issue.ts:29]: you can use `events.filter()`

File: src/helpers/issue.ts [hunk 7]
```diff
@@ -3,6 +3,37 @@ import { getBotConfig, getBotContext, getLogger } from "../bindings";
...
+    return events;
+  } catch (e: unknown) {
+    logger.debug(`Getting all issue events failed, reason: ${e}`);
+    return null;
+  }
+};
+
+export const getAllLabeledEvents = async () => {
+  const events = await getAllIssueEvents();
+  if (!events) return;
```
Comment (whilefoo) [inline src/helpers/issue.ts:27]:
```suggestion
  if (!events) return null;
```

File: src/helpers/issue.ts [hunk 8]
```diff
@@ -3,6 +3,37 @@ import { getBotConfig, getBotContext, getLogger } from "../bindings";
...
 import { checkRateLimitGit } from "../utils";
 
+export const getAllIssueEvents = async () => {
+  const context = getBotContext();
+  const logger = getLogger();
+  const payload = context.payload as Payload;
+  if (!payload.issue) return;
+  try {
+    // Fetch issue events
+    const { data: events } = await context.octokit.issues.listEvents({
```
Comment (whilefoo) [inline src/helpers/issue.ts:13]: you need to use pagination to retrieve all events, you can look at other functions as an example

**Review by whilefoo — CHANGES_REQUESTED** [review]
it works correctly, just needs some code improvements

File: src/helpers/issue.ts (hunk 8 above)
Comment (me505) [reply src/helpers/issue.ts:13]: whats pagination

Comment (me505) [reply src/helpers/issue.ts:13]: what shoul be page value

Comment (whilefoo) [reply src/helpers/issue.ts:13]: what do you mean?

Comment (me505) [reply src/helpers/issue.ts:13]: There was a per-page parameter and a page parameter, I just set the page parameter to one because idk

Comment (whilefoo) [reply src/helpers/issue.ts:13]: just make pagination like other functions

**Comment by Draeieg** [comment]
isnt this conflicting with #774? @pavlovcik

File: src/helpers/issue.ts (hunk 8 above)
Comment (whilefoo) [reply src/helpers/issue.ts:13]: example: [getAllIssueAssignEvents](https://github.com/ubiquity/ubiquibot/blob/5d10bda43e832717d7daba5949543e68f3dffbb8/src/helpers/issue.ts#L257)

Comment (me505) [reply src/helpers/issue.ts:13]: Sorry about it I was busy with school

**Comment by me505** [comment]
for such small things the bot should automatically use yarn lint --fix or somethig

File: src/helpers/issue.ts [hunk 9]
```diff
@@ -3,6 +3,52 @@ import { getBotConfig, getBotContext, getLogger } from "../bindings";
...
 
+export const getAllIssueEvents = async () => {
+  const context = getBotContext();
+  const logger = getLogger();
+  const payload = context.payload as Payload;
+  if (!payload.issue) return;
+
+  let shouldFetch = true;
+  let page_number = 1;
+  const events: any[] = [];
```
Comment (whilefoo) [inline src/helpers/issue.ts:14]:
```suggestion
  const events = [];
```
it will implicitly detect the type

**Review by whilefoo — CHANGES_REQUESTED** [review]
after you fix these, make QA again to verify it's working and we're good to go

**Comment by me505** [comment]
@whilefoo cant qa because of `Error: Cannot find module './lib/src/index.js' from '/workspaces/ubiquibot'`

**Comment by me505** [comment]
i tried yarn clean and then reinstall and start, didnt work

**Comment by whilefoo** [comment]
> @whilefoo cant qa because of `Error: Cannot find module './lib/src/index.js' from '/workspaces/ubiquibot'`

yes this is a known bug that's getting fixed. in the mean time you can change to 
```
"start": "probot run ./lib/index.js",
```
in your package.json

File: .github/ubiquibot-config.yml
Comment (pavlovcik) [inline .github/ubiquibot-config.yml:1]: You shouldn't be touching this file.

**Review by 0x4007 — CHANGES_REQUESTED** [review]

File: src/handlers/pricing/action.ts [hunk 10]
```diff
@@ -37,10 +37,51 @@ export const pricingLabelLogic = async (): Promise<void> => {
...
+    });
+
+    if (hasPriceLabel) {
+      // below checks each label to see if theres a price
+      if (hasTargetPriceLabel) {
+        // get all issue events of type "labeled" and the event label includes Price
+        const labeledEvents = await getAllLabeledEvents();
+        if (!labeledEvents) return;
+
+        labeledEvents.filter((event) => event.label?.name.includes("Price"));
```
Comment (whilefoo) [inline src/handlers/pricing/action.ts:63]: `filter` will return a new array so you need to store in the variable

**Review by whilefoo — CHANGES_REQUESTED** [review]

**Review by whilefoo — DISMISSED** [review]

**Comment by 0x4007** [comment]
Where's the proof that it works

**Comment by BeanieMen** [comment]
why does the bot hate me

**Comment by BeanieMen** [comment]
```
[Error: ENOENT: no such file or directory, open '/workspaces/ubiquibot/lib/assets/images/pmg.png'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/workspaces/ubiquibot/lib/assets/images/pmg.png'
}
error Command failed with exit code 1.
```

**Comment by BeanieMen** [comment]
theres no pmg.png in src/assets/images so i cant even jinx it

**Comment by whilefoo** [comment]
do this:
```
mkdir lib/assets/images
cp src/assets/images/flat.png lib/assets/images/
```

**Comment by cuddle-dev** [comment]
TLDR about the image path issue. 

We have been keeping the codebase for a long time which would not work with the lambda function in the telegram weekly post feature. If it's so panic to you @me505, you can comment it out and let me know if its being included in the merge into the development branch.

**Comment by BeanieMen** [comment]
@pavlovcik it prob can generate the permit for 25 usd not 400 as that whole thing wasnt modified

**Comment by BeanieMen** [comment]
@pavlovcik its generating correct amount now

**Comment by BeanieMen** [comment]
When will this get reviewed

**Comment by wannacfuture** [comment]
https://github.com/ubiquity/ubiquibot/pull/735#issuecomment-1736518788

Hey @me505  What about this?

**Comment by BeanieMen** [comment]
I haven't commented anything out

**Review by wannacfuture — DISMISSED** [review]

**Review by cuddle-dev — APPROVED** [review]
